### ISSUE TITLE
为什么不用qml呢

### ISSUE THREAD

**Issue by MidoriYakumo** [luspi/photoqt-dev#34]
既然都用Qt5了
C++内夹杂这大量css, 自定义UI看起来写得很复杂啊

**Comment by luspi** [comment]
If I understand it right (had to use google translator as I don't speak Chinese), you're wondering why I'm using CSS and not QML?

Well, the short answer to this is very simple: I don't know QML/QtQuick.

The slightly longer answer:

PhotoQt evolved more and more over time. As I started out I never thought that it'll grow in the way it has. I knew CSS pretty well, so that's what I started to use. By now, I do use quite a bit of CSS in PhotoQt, but I'm okay with it. I find CSS rather easy to understand. Plus, the code is broken down into many sub-classes, so I can re-use a lot of code over and over, so once I had it styled with CSS, I didn't have to worry about it anymore.

I had a look at porting the interface to QML/QtQuick. However, it's not quite a trivial task. It would take me a while to get a grip of QML/QtQuick and how to combine it with all the C++ stuff I'm doing in the background. And then I'd still have to port it over. Plus, the only real change would be that the whole styling thing looks differently, but I'm not sure that it'd necessarily be easier to understand.

Maybe I'm wrong. At some point (when I have some more time left) I do hope to learn more about QML/QtQuick, and maybe I change my opinion about it. But until then I'm rather happy with the way it is at the moment...

**Comment by alphachung** [comment]
@MidoriYakumo said: 
**Why don't you use QML?**
You already are using Qt5
Putting so many CSS inside C++, the custom UI code looks very complicated

**Comment by MidoriYakumo** [comment]
I would like to try to provide a migration from qss to qml for photoqt (a fork?) because this project is very nice.
I thought the developers may be Chinese so I used Chinese, sorry for that.

**Comment by luspi** [comment]
@alphachung 
Well, yes, I am using Qt5, however, the code is still based on QWidget. Using Qt5 doesn't make porting an easier task...

@MidoriYakumo 
You're more than welcome to look into that. However, be aware, that the interface is heavily interlinked with the rest of the code. So you might be more looking at a re-write than a port...


Well, doing some more reading about it, I might have another closer look at QML/QtQuick, too... I'll see if I can set up a basic image viewing proof-of-concept example. That will make it easier to see if and what advantages a full port would bring...

**Comment by luspi** [comment]
Okay, after spending a good few hours playing around with QML, here are my thoughts:

Doing the whole interface in QML would make it very easy to write the ui quickly. However, there is one problematic part I came across: The thumbnail bar.

QML has the (pretty amazing) possibility to specify imageproviders that are called automatically for images. However, even when using the exact same code I used before for loading images quickly, QML is significantly slower. Especially for larger directories it becomes a nightmare, as the slower performance becomes more and more apparent.

In the current QWidget implementation, PhotoQt first preloads the whole directory (depending on settings) with dummy thumbnails before loading the actual thumbnails as needed to keep the interface responsive. QML has the option to load images asynchronously, however, not only is it slower in loading images than my current QWidget implementation, an imagemodel doesn't display any images before all of them are loaded. There might be a way around this, but I wasn't able to find one.

A different approach would be, to keep the thumbnail bar as QGraphicsView and design the rest of the interface in QML using QQuickWidget. However, the problem that arises with that, is the transparency. QQuickWidget kinda destroys the transparency of the thumbnail bar (even more so, it somehow makes it completely transparent (including thumbnails and everything else) in such a way, that even manipulating the paintEvent of the thumbnailbar doesn't work anymore).

Both of the above ways unfortunately aren't a step forward for PhotoQt. They bring poor performance and don't integrate well with the rest of the application.

There is one possibility left that I haven't explored yet: It should be possible (so I read online) to expose the QGraphicsObject to QML and position them from there. I'll have a look at it, and if that's possible I'll certainly try to do that.

In conclusion: QML really seems to make life a lot easier for many of the things I'm doing with PhotoQt. However, so far, there are still some limitations that disqualify it for PhotoQt -> I wont sacrifice critical performance for easier code! I'll have a look at the last possibility I mentioned above, but it might just be that QML is not (yet) compatible to what PhotoQt requires of it...

**Comment by MidoriYakumo** [comment]
Have you profiled the performance of that piece of qml code? I think QML Profiler may help you.

**Comment by Harinlen** [comment]
I have been read some of your codes in mainwindow.cpp, it seems that some stylesheet can be replaced by setPalette().
e.g.:
You have the following code(#27 mainwindow.cpp):

        this->setStyleSheet("#mainwindow { background-color: rgba(0,0,0,0); }");

It seems that you can use the following code instead:

	QPalette pal=palette();
	pal.setPalette(QPalette::Base, QColor(0,0,0,0));
	pal.setPalette(QPalette::Window, QColor(0,0,0,0));
	setPalette(pal);

Someone tell me that we'd better not use style sheet in C++ codes mixed in native style of any platform except using the official "fusion" style. 
But for some UI features, style sheet is the simpliest way, like the border and menu seperator color. Even so, I think using QPalette is much better than using style sheet. You can configure all the palette with a class and configure file. And load them according to the object name, it's swift and easy.
P.S.: Amazing project. I like photoqt very much!

**Comment by luspi** [comment]
@MidoriYakumo 
I haven't done that yet, but I'll have a look at it, maybe I can figure out what's causing the slowness...

@Harinlen 
Well, what I've heard is that one just shouldn't mix QPalette and stylesheet usage, but I might be wrong there...
For quite a few of the things I'm doing, I'm pretty sure I have to use stylesheets; e.g., for styling a QSlider, with stylesheets I can easily set a custom look to all the different elements of the QSlider (handler active&inactive, slider active&inactive, background). As to my knowledge, this isn't possible with QPalette (tell me if I'm wrong).
However, being able to configure if not all, maybe most (or at least some) of the colors and appearances of PhotoQt using a simple configuration file sounds pretty cool actually. I'll look into that...

What I'm doing right now: I'm trying to implement proper thumbnail support in QML. If I move as much of the computations as possible to the C++ backend, it might be possible to have an all-QML gui that is as fast and responsive as the current QWidget interface. However, it does require a lot of changes to the logic I used for creating thumbnails, so it might take a little while. But I'm up for giving it a go. And if in the end it doesn't seem to be possible and I decide to not go this way, at least I'll get some experience with QML ;-)

**Comment by luspi** [comment]
Okay, so, I've been playing around some more, and I actually got a basic version kinda running (though it has almost no features yet except thumbnails and, of course, the main image). I'm still tweaking at the thumbnails, they aren't yet fully working, but it seems to be possible (when doing all the heavy computations in C++)... Once I got them working properly, I'll probably create a branch here with the QML code and start porting/adding the rest...

I wont go as far as to promise that a QML version is coming (it all depends on the thumbnails, as mentioned in previous comments), but it does seem rather likely...

I have initially been rather hesitant towards QML, but the more I get to know it - it really makes life much easier! It's very handy to seperate view and computations a little more! I don't think that QML is the best solution for any app, but it certainly seems (more and more) like it is a good solution for PhotoQt...

**Comment by luspi** [comment]
I have some good news: I got a pretty solid and efficient implementation of the whole thumbnails thing in C++/QML (I made some changes, so I think it is even slightly faster now than it has been in v1.2). So, I've created a new repository - https://github.com/luspi/photoqt-qml-dev - where I'll be developing a QML port of PhotoQt... Thanks for hinting me to QML, guys! I really have been hesitant for no reason...

Note, that PhotoQt-QML is still veerrrry basic, but at least the thumbnails are working pretty well already ;-)

**Comment by luspi** [comment]
PhotoQt is moving away from GitHub to GitLab ([more info](http://photoqt.org/news/1434375887)). This issue is duplicated here: https://gitlab.com/luspi/photoqt-legacy/issues
